cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,192 @@
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &ParallelizeArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let mut tx = workspace_command.start_transaction();
+    // The target commits are the commits being parallelized.
+    let target_commits: Vec<CommitId> = tx
+        .base_workspace_helper()
+        .parse_union_revsets(&args.revisions)?
+        .evaluate()?
+        .iter()
+        // We want parents before children, so we need to reverse the order.
+        .reversed()
+        .collect();
+    if target_commits.len() < 2 {
+        return Ok(writeln!(ui.status(), "Nothing changed.")?);
+    }
+    let target_revset = RevsetExpression::commits(target_commits.clone());
+    let target_head = get_head(&target_revset, tx.repo())?;
+    let target_root = get_root(&target_revset, tx.repo())?;
+    let connected_length = RevsetExpression::commits(vec![target_head, target_root.clone()])
+        .connected()
+        .evaluate_programmatic(tx.repo())?
+        .iter()
+        .count();
+    if connected_length != target_commits.len() {
+        return Err(user_error(
+            "Cannot parallelize since the target revisions are not connected.",
+        ));
+    }
+
+    // Rebase the children of each target commit onto its new parents. A child
+    // which had a target commit as an ancestor before parallelize ran will have
+    // the target commit as a parent afterward.
+    for target_commit in target_commits.iter() {
+        // These parents are shared by all children of the target commit and
+        // include the target commit itself plus any of its ancestors which are
+        // being parallelized.
+        let common_parents: IndexSet<CommitId> = target_revset
+            .intersection(&RevsetExpression::commit(target_commit.clone()).ancestors())
+            .evaluate_programmatic(tx.repo())?
+            .iter()
+            // We want parents before children, so we need to reverse the order.
+            .reversed()
+            .collect();
+        // Children of the target commit, excluding other target commits.
+        let children: Vec<Commit> = RevsetExpression::commit(target_commit.clone())
+            .children()
+            .minus(&target_revset)
+            .evaluate_programmatic(tx.repo())?
+            .iter()
+            .commits(tx.repo().store())
+            .try_collect()?;
+        for child in children {
+            let mut new_parents = common_parents.clone();
+            new_parents.extend(child.parent_ids().iter().cloned());
+            let parents = new_parents
+                .iter()
+                .map(|c| tx.repo().store().get_commit(c))
+                .try_collect::<_, Vec<Commit>, _>()?;
+            rebase_descendants(
+                &mut tx,
+                command.settings(),
+                &parents,
+                &[child],
+                Default::default(),
+            )?;
+        }
```

Comment:
This is a bit inefficient because it looks for children every time (children are a bit expensive to find because there's no index of parent->children, so we find them by walking from the visible heads). I think it's fine for now. It should be easier to simplify it when I finish my rebase API improvements.

Reply:
Would it be more efficient to get the children of the entire target revset in one shot and find which commits are their parents using the backlinks? I guess that would be a lot more code so maybe it's not worth it.